File: 2018_kakao_blind_recruit_cache.py
def solution(cacheSize, cities):
    answer = 0
    cache = {}
    for city in cities:
        city = city.lower()
        if cacheSize == 0:
            answer+=5
            continue
           
        for c in cache: cache[c] += 1
            
        if city in cache:
            cache[city] = 0
            answer += 1
        else:
            answer += 5
            if len(cache) < cacheSize:
                cache[city] = 0
            else:
                lru = sorted(cache.keys(), key=lambda x: -cache[x])[0]
                del cache[lru]
                cache[city] = 0
    
    return answer

File: 2018_kakao_blind_recruit_cache.js
function solution(cacheSize, cities) {
  if (cacheSize === 0) return cities.length * 5;

  let answer = 0;
  const cache = {};
  let curSize = 0;

  cities.forEach((cityName) => {
    const city = cityName.toLowerCase();

    Object.keys(cache).forEach((c) => (cache[c] += 1));

    if (city in cache) {
      cache[city] = 0;
      answer += 1;
    } else {
      answer += 5;
      if (curSize < cacheSize) {
        curSize += 1;
      } else {
        const maxCache = Object.keys(cache).sort(
          (a, b) => cache[b] - cache[a]
        )[0];
        delete cache[maxCache];
      }
      cache[city] = 0;
    }
  });

  return answer;
}
